src/Interpreters/InterpreterCheckQuery.cpp

```diff
@@ -8,72 +8,201 @@
+        {names_and_types[1].type->createColumn(), names_and_types[1].type, names_and_types[1].name},
+        {names_and_types[2].type->createColumn(), names_and_types[2].type, names_and_types[2].name},
+    });
 }
 
+Chunk getChunkFromCheckResult(const CheckResult & check_result)
+{
+    MutableColumns columns = getHeaderForCheckResult().cloneEmptyColumns();
+    columns[0]->insert(check_result.fs_path);
+    columns[1]->insert(static_cast<UInt8>(check_result.success));
+    columns[2]->insert(check_result.failure_message);
+    return Chunk(std::move(columns), 1);
 }
 
+class TableCheckWorkerProcessor : public ISource
+{
+
+public:
+    TableCheckWorkerProcessor(IStorage::DataValidationTasksPtr check_data_tasks_, StoragePtr table_)
+        : ISource(getHeaderForCheckResult())
+        , table(table_)
+        , check_data_tasks(check_data_tasks_)
+    {
+    }
+
+    String getName() const override { return "TableCheckWorkerProcessor"; }
 
-InterpreterCheckQuery::InterpreterCheckQuery(const ASTPtr & query_ptr_, ContextPtr context_) : WithContext(context_), query_ptr(query_ptr_)
+protected:
+
+    std::optional<Chunk> tryGenerate() override
+    {
+        bool has_nothing_to_do = false;
+        auto check_result = table->checkDataNext(check_data_tasks, has_nothing_to_do);
+        if (has_nothing_to_do)
+            return {};
+
+        /// We can omit manual `progess` call, ISource will may count it automatically by returned chunk
+        /// However, we want to report only rows in progress
+        progress(1, 0);
+
+        if (!check_result.success)
+        {
+            LOG_WARNING(&Poco::Logger::get("InterpreterCheckQuery"),
+                "Check query for table {} failed, path {}, reason: {}",
+                table->getStorageID().getNameForLogs(),
+                check_result.fs_path,
+                check_result.failure_message);
+        }
+
+        return getChunkFromCheckResult(check_result);
+    }
+
+private:
+    StoragePtr table;
+    IStorage::DataValidationTasksPtr check_data_tasks;
+};
+
+class TableCheckResultEmitter : public IAccumulatingTransform
 {
+public:
+    TableCheckResultEmitter() : IAccumulatingTransform(getHeaderForCheckResult(), getSingleValueBlock(1).cloneEmpty()) {}
+
+    String getName() const override { return "TableCheckResultEmitter"; }
+
+    void consume(Chunk chunk) override
+    {
+        if (result_value == 0)
+            return;
+
+        auto columns = chunk.getColumns();
+        if (columns.size() != 3)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Wrong number of columns: {}", columns.size());
+
+        const auto * col = checkAndGetColumn<ColumnUInt8>(columns[1].get());
+        for (size_t i = 0; i < col->size(); ++i)
+        {
+            if (col->getElement(i) == 0)
+            {
+                result_value = 0;
+                return;
+            }
+        }
+    }
+
+    Chunk generate() override
+    {
+        if (is_valuer_emitted.exchange(true))
```

Comment:
we return chunk only once, right?

Reply:
Yes, only one block with single row is expected in the result